snuba/replacers/errors_replacer.py

```diff
@@ -478,50 +475,103 @@ def get_timestamp_condition(msg_field: str, operator: str) -> str:
 
     query_args = {
         "event_ids": event_id_list,
-        "project_id": message["project_id"],
+        "project_id": str(project_id),
     }
 
     return prewhere, where, query_args
 
 
-def process_replace_group(
-    message: ReplacementMessage[ReplaceGroupMessageBody],
-    all_columns: Sequence[FlattenedColumn],
-    state_name: ReplacerState,
-) -> Optional[Replacement]:
+@dataclass
+class ReplaceGroupReplacement(Replacement):
     """
     Merge individual events into new group. The old group will have to be
     manually excluded from search queries.
 
-    See docstring of process_exclude_groups for an explanation of how this is used.
+    See docstring of ExcludeGroupsReplacement for an explanation of how this is
+    used.
 
     Note that events merged this way cannot be cleanly unmerged by
     process_unmerge, as their group hashes possibly stand in no correlation to
     how the merging was done.
     """
 
-    event_ids = message.data["event_ids"]
-    if not event_ids:
-        return None
+    event_ids: Sequence[str]
+    project_id: int
+    from_timestamp: Optional[str]
+    to_timestamp: Optional[str]
+    new_group_id: int
+    replacement_type: ReplacementType
+    replacement_message_metadata: ReplacementMessageMetadata
+    all_columns: Sequence[FlattenedColumn]
 
-    event_set_filter = _build_event_set_filter(message.data, state_name)
-    if event_set_filter is None:
+    @classmethod
+    def parse_message(
+        cls,
+        message: ReplacementMessage[ReplaceGroupMessageBody],
+        context: ReplacementContext,
+    ) -> Optional[ReplaceGroupReplacement]:
+        event_ids = message.data["event_ids"]
+        if not event_ids:
+            return None
+
+        return cls(
+            event_ids=event_ids,
+            project_id=message.data["project_id"],
+            from_timestamp=message.data.get("from_timestamp"),
+            to_timestamp=message.data.get("to_timestamp"),
+            new_group_id=message.data["new_group_id"],
+            replacement_type=message.action_type,
+            replacement_message_metadata=message.metadata,
+            all_columns=context.all_columns,
+        )
+
+    def get_project_id(self) -> int:
+        return self.project_id
+
+    def get_query_time_flags(self) -> Optional[QueryTimeFlags]:
         return None
 
-    prewhere, where, query_args = event_set_filter
+    def get_replacement_type(self) -> ReplacementType:
+        return self.replacement_type
 
-    full_where = f"PREWHERE {' AND '.join(prewhere)} WHERE {' AND '.join(where)}"
-    project_id: int = message.data["project_id"]
-    query_time_flags = (None, project_id)
+    def get_message_metadata(self) -> ReplacementMessageMetadata:
+        return self.replacement_message_metadata
 
-    return _build_group_replacement(
```

Comment:
_build_group_replacement was a bad abstraction that got in the way of refactoring, so I just inlined the relevant bits.